src/ringbuffer.rs

```diff
@@ -0,0 +1,369 @@
+	M: StorageMap<Index, I, Query = I>,
+	T: RingBufferTrait<I, Bounds = B, Map = M> + ?Sized,
+{
+	start: Index,
+	end: Index,
+	_t: PhantomData<(I, B, M, T)>,
+}
+
+impl<I, B, M, T> RingBufferTransient<I, B, M, T>
+where
+	I: Codec + EncodeLike,
+	B: StorageValue<(Index, Index), Query = (Index, Index)>,
+	M: StorageMap<Index, I, Query = I>,
+	T: RingBufferTrait<I, Bounds = B, Map = M> + ?Sized,
+{
+	/// Create a new RingBufferTransient that backs the ringbuffer implementation.
+	///
+	/// Initializes itself from the Bounds.
+	pub fn new() -> RingBufferTransient<I, B, M, T> {
+		let (start, end) = <<T as RingBufferTrait<I>>::Bounds>::get();
+		RingBufferTransient {
+			start,
+			end,
+			_t: PhantomData,
+		}
+	}
+}
+
+impl<I, B, M, T> Drop for RingBufferTransient<I, B, M, T>
+where
+	I: Codec + EncodeLike,
+	B: StorageValue<(Index, Index), Query = (Index, Index)>,
+	M: StorageMap<Index, I, Query = I>,
+	T: RingBufferTrait<I, Bounds = B, Map = M> + ?Sized,
+{
+	fn drop(&mut self) {
+		<Self as RingBufferTrait<I>>::commit(self);
+	}
+}
+
+/// Trait object presenting the ringbuffer interface.
+pub trait RingBufferTrait<I>
+where
+	I: Codec + EncodeLike,
+{
+	type Bounds: StorageValue<(Index, Index)>;
+	type Map: StorageMap<Index, I>;
+
+	/// Store all changes made in the underlying storage.
+	///
+	/// Data is not guaranteed to be consistent before this call.
+	/// Will be called by `drop`.
+	fn commit(&self);
+
+	/// Push an item onto the end of the queue.
+	fn push(&mut self, i: I);
+	/// Push an item onto the front of the queue.
+	fn push_front(&mut self, i: I);
+
+	/// Pop an item from the start of the queue.
+	///
+	/// Returns `None` if the queue is empty.
+	fn pop(&mut self) -> Option<I>;
+}
+
+/// Ringbuffer implementation based on `RingBufferTransient`
+impl<I, B, M, T> RingBufferTrait<I> for RingBufferTransient<I, B, M, T>
+where
+	I: Codec + EncodeLike,
+	B: StorageValue<(Index, Index), Query = (Index, Index)>,
+	M: StorageMap<Index, I, Query = I>,
+	T: RingBufferTrait<I, Bounds = B, Map = M> + ?Sized,
+{
+	type Bounds = B;
+	type Map = M;
+
+	/// Commit the (potentially) changed bounds to storage.
+	fn commit(&self) {
+		Self::Bounds::put((self.start, self.end));
+	}
+
+	/// Push an item onto the end of the queue.
+	///
+	/// Will insert the new item, but will not update the bounds in storage.
+	fn push(&mut self, item: I) {
+		Self::Map::insert(self.end, item);
+		// this will intentionally overflow and wrap around when bonds_end
+		// reaches `Index::max_value` because we want a ringbuffer.
+		let next_index = self.end.wrapping_add(1);
+		if next_index == self.start {
+			// overwrite the oldest item in the FIFO ringbuffer
+			self.start = self.start.wrapping_add(1);
+			self.end = next_index;
+		} else {
+			self.end = next_index;
+		}
+	}
+
+	/// Push an item onto the front of the queue.
+	///
+	/// Equivalent to `push` if the queue is empty.
+	fn push_front(&mut self, item: I) {
+		if self.start == self.end {
```

Comment:
Are you sure this is expected behavior?

Say `start = 2`, `end = 1` (so basically I've been calling `push_front()` repeatedly since the beginning, and start went from 0 -> max -> max - 1 ... -> 3 -> 2).
I call `push_front` again, getting me to `start == 1` and `start == end`.
Now calling `push_front` again would get us to: `start == 1` and `end == 2`, when in my mind it should be `end == 0` and `start == 1`.

Is my reasoning wrong?

Reply:
Do you think this is addressed by keeping up the "`start == end` means `is_empty`" invariant?

Reply:
Not sure what you mean here: the changes you made look good, only problem I see remaining is the `is_empty()` function but I'm sure you already know that ^^